File: src/SIL.Machine.TS/src/translation/error-correction-word-graph-processor.ts
import { PriorityQueue } from 'typescript-collections';

import { LOG_ZERO } from '../statistics/log-space';
import { EcmScoreInfo } from './ecm-score-info';
import { ErrorCorrectionModel } from './error-correction-model';
import { TranslationResult } from './translation-result';
import { TranslationResultBuilder } from './translation-result-builder';
import { WordAlignmentMatrix } from './word-alignment-matrix';
import { INITIAL_STATE, WordGraph } from './word-graph';
import { WordGraphArc } from './word-graph-arc';

export class ErrorCorrectionWordGraphProcessor {
  confidenceThreshold: number = 0;

  private readonly restScores: number[];
  private readonly stateEcmScoreInfos: EcmScoreInfo[] = [];
  private readonly arcEcmScoreInfos: EcmScoreInfo[][] = [];
  private readonly stateBestScores: number[][] = [];
  private readonly stateWordGraphScores: number[] = [];
  private readonly stateBestPrevArcs: number[][] = [];
  private readonly statesInvolvedInArcs: Set<number> = new Set<number>();
  private prevPrefix: string[] = [];
  private prevIsLastWordComplete: boolean = false;

  constructor(
    private readonly ecm: ErrorCorrectionModel,
    private readonly sourceSegment: string[],
    private readonly wordGraph: WordGraph,
    public readonly ecmWeight = 1,
    public readonly wordGraphWeight = 1
  ) {
    this.restScores = this.wordGraph.computeRestScores();

    this.initStates();
    this.initArcs();
  }

  correct(prefix: string[], isLastWordComplete: boolean, n: number): TranslationResult[] {
    // get valid portion of the processed prefix vector
    let validProcPrefixCount = 0;
    for (let i = 0; i < this.prevPrefix.length; i++) {
      if (i >= prefix.length) {
        break;
      }

      if (i === this.prevPrefix.length - 1 && i === prefix.length - 1) {
        if (this.prevPrefix[i] === prefix[i] && this.prevIsLastWordComplete === isLastWordComplete) {
          validProcPrefixCount++;
        }
      } else if (this.prevPrefix[i] === prefix[i]) {
        validProcPrefixCount++;
      }
    }

    const diffSize = this.prevPrefix.length - validProcPrefixCount;
    if (diffSize > 0) {
      // adjust size of info for arcs
      for (const esis of this.arcEcmScoreInfos) {
        for (const esi of esis) {
          for (let i = 0; i < diffSize; i++) {
            esi.removeLast();
          }
        }
      }

      // adjust size of info for states
      for (const state of this.statesInvolvedInArcs) {
        for (let i = 0; i < diffSize; i++) {
          this.stateEcmScoreInfos[state].removeLast();
          this.stateBestScores[state].pop();
          this.stateBestPrevArcs[state].pop();
        }
      }
    }

    // get difference between prefix and valid portion of processed prefix
    const prefixDiff = new Array<string>(prefix.length - validProcPrefixCount);
    for (let i = 0; i < prefixDiff.length; i++) {
      prefixDiff[i] = prefix[validProcPrefixCount + i];
    }

    // process word-graph given prefix difference
    this.processWordGraphForPrefixDiff(prefixDiff, isLastWordComplete);

    this.prevPrefix = prefix;
    this.prevIsLastWordComplete = isLastWordComplete;

    const queue = new PriorityQueue<Hypothesis>((x, y) => {
      if (x.score < y.score) {
        return -1;
      } else if (x.score > y.score) {
        return 1;
      } else {
        return 0;
      }
    });
    this.getStateHypotheses(queue);
    this.getSubStateHypotheses(queue);

    const results: TranslationResult[] = [];
    for (const hypothesis of this.nbestSearch(n, queue)) {
      const builder = new TranslationResultBuilder();
      this.buildCorrectionFromHypothesis(builder, prefix, isLastWordComplete, hypothesis);
      results.push(builder.toResult(this.sourceSegment, prefix.length));
    }
    return results;
  }

  private initStates(): void {
    for (let i = 0; i < this.wordGraph.stateCount; i++) {
      this.stateEcmScoreInfos.push(new EcmScoreInfo());
      this.stateWordGraphScores.push(LOG_ZERO);
      this.stateBestScores.push([]);
      this.stateBestPrevArcs.push([]);
    }

    if (!this.wordGraph.isEmpty) {
      this.ecm.setupInitialEsi(this.stateEcmScoreInfos[INITIAL_STATE]);
      this.updateInitialStateBestScores();
    }
  }

  private initArcs(): void {
    for (let arcIndex = 0; arcIndex < this.wordGraph.arcs.length; arcIndex++) {
      const arc = this.wordGraph.arcs[arcIndex];

      // init ecm score info for each word of arc
      let prevEsi = this.stateEcmScoreInfos[arc.prevState];
      const esis: EcmScoreInfo[] = [];
      for (const word of arc.words) {
        const esi = new EcmScoreInfo();
        this.ecm.setupEsi(esi, prevEsi, word);
        esis.push(esi);
        prevEsi = esi;
      }
      this.arcEcmScoreInfos.push(esis);

      // init best scores for the arc's successive state
      this.updateStateBestScores(arcIndex, 0);

      this.statesInvolvedInArcs.add(arc.prevState);
      this.statesInvolvedInArcs.add(arc.nextState);
    }
  }

  private updateInitialStateBestScores(): void {
    const esi = this.stateEcmScoreInfos[INITIAL_STATE];

    this.stateWordGraphScores[INITIAL_STATE] = this.wordGraph.initialStateScore;

    const bestScores = this.stateBestScores[INITIAL_STATE];
    const bestPrevArcs = this.stateBestPrevArcs[INITIAL_STATE];

    bestScores.length = 0;
    bestPrevArcs.length = 0;

    for (const score of esi.scores) {
      bestScores.push(this.ecmWeight * -score + this.wordGraphWeight * this.wordGraph.initialStateScore);
      bestPrevArcs.push(Number.MAX_SAFE_INTEGER);
    }
  }

  private updateStateBestScores(arcIndex: number, prefixDiffSize: number): void {
    const arc = this.wordGraph.arcs[arcIndex];
    const arcEsis = this.arcEcmScoreInfos[arcIndex];

    const prevEsi = arcEsis.length === 0 ? this.stateEcmScoreInfos[arc.prevState] : arcEsis[arcEsis.length - 1];

    const wordGraphScore = this.stateWordGraphScores[arc.prevState] + arc.score;

    const nextStateBestScores = this.stateBestScores[arc.nextState];
    const nextStateBestPrevArcs = this.stateBestPrevArcs[arc.nextState];

    const positions: number[] = [];
    const startPos = prefixDiffSize === 0 ? 0 : prevEsi.scores.length - prefixDiffSize;
    for (let i = startPos; i < prevEsi.scores.length; i++) {
      const newScore = this.ecmWeight * -prevEsi.scores[i] + this.wordGraphWeight * wordGraphScore;

      if (i === nextStateBestScores.length || nextStateBestScores[i] < newScore) {
        this.addOrReplace(nextStateBestScores, i, newScore);
        positions.push(i);
        this.addOrReplace(nextStateBestPrevArcs, i, arcIndex);
      }
    }

    this.stateEcmScoreInfos[arc.nextState].updatePositions(prevEsi, positions);

    if (wordGraphScore > this.stateWordGraphScores[arc.nextState]) {
      this.stateWordGraphScores[arc.nextState] = wordGraphScore;
    }
  }

  private addOrReplace<T>(list: T[], index: number, item: T): void {
    if (index > list.length) {
      throw new Error('Index is out of range.');
    }

    if (index === list.length) {
      list.push(item);
    } else {
      list[index] = item;
    }
  }

  private processWordGraphForPrefixDiff(prefixDiff: string[], isLastWordComplete: boolean): void {
    if (prefixDiff.length === 0) {
      return;
    }

    if (!this.wordGraph.isEmpty) {
      const prevInitialEsi = this.stateEcmScoreInfos[INITIAL_STATE];
      this.ecm.extendInitialEsi(this.stateEcmScoreInfos[INITIAL_STATE], prevInitialEsi, prefixDiff);
      this.updateInitialStateBestScores();
    }

    for (let arcIndex = 0; arcIndex < this.wordGraph.arcs.length; arcIndex++) {
      const arc = this.wordGraph.arcs[arcIndex];

      // update ecm score info for each word of arc
      let prevEsi = this.stateEcmScoreInfos[arc.prevState];
      const esis = this.arcEcmScoreInfos[arcIndex];
      while (esis.length < arc.words.length) {
        esis.push(new EcmScoreInfo());
      }
      for (let i = 0; i < arc.words.length; i++) {
        const esi = esis[i];
        this.ecm.extendEsi(esi, prevEsi, arc.isUnknown ? '' : arc.words[i], prefixDiff, isLastWordComplete);
        prevEsi = esi;
      }

      // update best scores for the arc's successive state
      this.updateStateBestScores(arcIndex, prefixDiff.length);
    }
  }

  private getStateHypotheses(queue: PriorityQueue<Hypothesis>): void {
    for (const state of this.statesInvolvedInArcs) {
      const restScore = this.restScores[state];
      const bestScores = this.stateBestScores[state];

      const score = bestScores[bestScores.length - 1] + this.wordGraphWeight * restScore;
      queue.enqueue(new Hypothesis(score, state));
    }
  }

  private getSubStateHypotheses(queue: PriorityQueue<Hypothesis>): void {
    for (let arcIndex = 0; arcIndex < this.wordGraph.arcs.length; arcIndex++) {
      const arc = this.wordGraph.arcs[arcIndex];
      if (arc.words.length > 1 && !this.isArcPruned(arc)) {
        const wordGraphScore = this.stateWordGraphScores[arc.prevState] + arc.score;

        for (let i = 0; i < arc.words.length; i++) {
          const esi = this.arcEcmScoreInfos[arcIndex][i];
          const score =
            this.wordGraphWeight * wordGraphScore +
            this.ecmWeight * -esi.scores[esi.scores.length - 1] +
            this.wordGraphWeight * this.restScores[arc.nextState];
          queue.enqueue(new Hypothesis(score, arc.nextState, arcIndex, i));
        }
      }
    }
  }

  private isArcPruned(arc: WordGraphArc): boolean {
    return !arc.isUnknown && arc.wordConfidences.some(c => c < this.confidenceThreshold);
  }

  private nbestSearch(n: number, queue: PriorityQueue<Hypothesis>): Hypothesis[] {
    const nbest: Hypothesis[] = [];
    while (!queue.isEmpty()) {
      const hypothesis = queue.dequeue();
      if (hypothesis == null) {
        break;
      }
      const lastState =
        hypothesis.arcs.length === 0 ? hypothesis.startState : hypothesis.arcs[hypothesis.arcs.length - 1].nextState;

      if (this.wordGraph.finalStates.has(lastState)) {
        nbest.push(hypothesis);
        if (nbest.length === n) {
          break;
        }
      } else if (this.confidenceThreshold <= 0) {
        hypothesis.arcs.push(...this.wordGraph.getBestPathFromStateToFinalState(lastState));
        nbest.push(hypothesis);
        if (nbest.length === n) {
          break;
        }
      } else {
        const score = hypothesis.score - this.wordGraphWeight * this.restScores[lastState];
        const arcIndices = this.wordGraph.getNextArcIndices(lastState);
        let enqueuedArc = false;
        for (let i = 0; i < arcIndices.length; i++) {
          const arcIndex = arcIndices[i];
          const arc = this.wordGraph.arcs[arcIndex];
          if (this.isArcPruned(arc)) {
            continue;
          }

          let newHypothesis = hypothesis;
          if (i < arcIndices.length - 1) {
            newHypothesis = newHypothesis.clone();
          }
          newHypothesis.score = score;
          newHypothesis.score += arc.score;
          newHypothesis.score += this.restScores[arc.nextState];
          newHypothesis.arcs.push(arc);
          queue.enqueue(newHypothesis);
          enqueuedArc = true;
        }

        if (!enqueuedArc && (hypothesis.startArcIndex !== -1 || hypothesis.arcs.length > 0)) {
          hypothesis.arcs.push(...this.wordGraph.getBestPathFromStateToFinalState(lastState));
          nbest.push(hypothesis);
          if (nbest.length === n) {
            break;
          }
        }
      }
    }
    return nbest;
  }

  private buildCorrectionFromHypothesis(
    builder: TranslationResultBuilder,
    prefix: string[],
    isLastWordComplete: boolean,
    hypothesis: Hypothesis
  ): void {
    let uncorrectedPrefixLen: number;
    if (hypothesis.startArcIndex === -1) {
      this.addBestUncorrectedPrefixState(builder, prefix.length, hypothesis.startState);
      uncorrectedPrefixLen = builder.words.length;
    } else {
      this.addBestUncorrectedPrefixSubState(
        builder,
        prefix.length,
        hypothesis.startArcIndex,
        hypothesis.startArcWordIndex
      );
      const firstArc = this.wordGraph.arcs[hypothesis.startArcIndex];
      uncorrectedPrefixLen = builder.words.length - (firstArc.words.length - hypothesis.startArcWordIndex) + 1;
    }

    let alignmentColsToAddCount = this.ecm.correctPrefix(builder, uncorrectedPrefixLen, prefix, isLastWordComplete);

    for (const arc of hypothesis.arcs) {
      this.updateCorrectionFromArc(builder, arc, false, alignmentColsToAddCount);
      alignmentColsToAddCount = 0;
    }
  }

  private addBestUncorrectedPrefixState(builder: TranslationResultBuilder, procPrefixPos: number, state: number): void {
    const arcs: WordGraphArc[] = [];

    let curState = state;
    let curProcPrefixPos = procPrefixPos;
    while (curState !== 0) {
      const arcIndex = this.stateBestPrevArcs[curState][curProcPrefixPos];
      const arc = this.wordGraph.arcs[arcIndex];

      for (let i = arc.words.length - 1; i >= 0; i--) {
        const predPrefixWords = this.arcEcmScoreInfos[arcIndex][i].getLastInsPrefixWordFromEsi();
        curProcPrefixPos = predPrefixWords[curProcPrefixPos];
      }

      arcs.unshift(arc);

      curState = arc.prevState;
    }

    for (const arc of arcs) {
      this.updateCorrectionFromArc(builder, arc, true, 0);
    }
  }

  private addBestUncorrectedPrefixSubState(
    builder: TranslationResultBuilder,
    procPrefixPos: number,
    arcIndex: number,
    arcWordIndex: number
  ): void {
    const arc = this.wordGraph.arcs[arcIndex];

    let curProcPrefixPos = procPrefixPos;
    for (let i = arcWordIndex; i >= 0; i--) {
      const predPrefixWords = this.arcEcmScoreInfos[arcIndex][i].getLastInsPrefixWordFromEsi();
      curProcPrefixPos = predPrefixWords[curProcPrefixPos];
    }

    this.addBestUncorrectedPrefixState(builder, curProcPrefixPos, arc.prevState);

    this.updateCorrectionFromArc(builder, arc, true, 0);
  }

  private updateCorrectionFromArc(
    builder: TranslationResultBuilder,
    arc: WordGraphArc,
    isPrefix: boolean,
    alignmentColsToAddCount: number
  ): void {
    for (let i = 0; i < arc.words.length; i++) {
      builder.appendWord(arc.words[i], arc.wordConfidences[i], !isPrefix && arc.isUnknown);
    }

    let alignment = arc.alignment;
    if (alignmentColsToAddCount > 0) {
      const newAlignment = new WordAlignmentMatrix(alignment.rowCount, alignment.columnCount + alignmentColsToAddCount);
      for (let j = 0; j < alignment.columnCount; j++) {
        for (let i = 0; i < alignment.rowCount; i++) {
          newAlignment.set(i, alignmentColsToAddCount + j, alignment.get(i, j));
        }
      }
      alignment = newAlignment;
    }

    builder.markPhrase(arc.sourceSegmentRange, alignment);
  }
}

class Hypothesis {
  readonly arcs: WordGraphArc[] = [];

  constructor(
    public score: number,
    public readonly startState: number,
    public readonly startArcIndex: number = -1,
    public readonly startArcWordIndex = -1
  ) {}

  clone(): Hypothesis {
    const clone = new Hypothesis(this.score, this.startState, this.startArcIndex, this.startArcWordIndex);
    clone.arcs.push(...this.arcs);
    return clone;
  }
}

File: src/SIL.Machine/Translation/ErrorCorrectionWordGraphProcessor.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using SIL.Machine.DataStructures;
using SIL.Machine.Statistics;
using SIL.Machine.Tokenization;

namespace SIL.Machine.Translation
{
    public class ErrorCorrectionWordGraphProcessor
    {
        private readonly WordGraph _wordGraph;
        private readonly double[] _restScores;
        private readonly ErrorCorrectionModel _ecm;
        private readonly IDetokenizer<string, string> _targetDetokenizer;
        private readonly List<EcmScoreInfo> _stateEcmScoreInfos;
        private readonly List<List<EcmScoreInfo>> _arcEcmScoreInfos;
        private readonly List<List<double>> _stateBestScores;
        private readonly List<double> _stateWordGraphScores;
        private readonly List<List<int>> _stateBestPrevArcs;
        private readonly HashSet<int> _statesInvolvedInArcs;
        private string[] _prevPrefix;
        private bool _prevIsLastWordComplete;

        public ErrorCorrectionWordGraphProcessor(
            ErrorCorrectionModel ecm,
            IDetokenizer<string, string> targetDetokenizer,
            WordGraph wordGraph,
            double ecmWeight = 1,
            double wordGraphWeight = 1
        )
        {
            _ecm = ecm;
            _targetDetokenizer = targetDetokenizer;
            _wordGraph = wordGraph;
            EcmWeight = ecmWeight;
            WordGraphWeight = wordGraphWeight;

            _restScores = _wordGraph.ComputeRestScores().ToArray();
            _stateEcmScoreInfos = new List<EcmScoreInfo>();
            _arcEcmScoreInfos = new List<List<EcmScoreInfo>>();
            _stateBestScores = new List<List<double>>();
            _stateWordGraphScores = new List<double>();
            _stateBestPrevArcs = new List<List<int>>();
            _statesInvolvedInArcs = new HashSet<int>();
            _prevPrefix = new string[0];

            InitStates();
            InitArcs();
        }

        public double ConfidenceThreshold { get; set; }

        private void InitStates()
        {
            for (int i = 0; i < _wordGraph.StateCount; i++)
            {
                _stateEcmScoreInfos.Add(new EcmScoreInfo());
                _stateWordGraphScores.Add(LogSpace.Zero);
                _stateBestScores.Add(new List<double>());
                _stateBestPrevArcs.Add(new List<int>());
            }

            if (!_wordGraph.IsEmpty)
            {
                _ecm.SetupInitialEsi(_stateEcmScoreInfos[WordGraph.InitialState]);
                UpdateInitialStateBestScores();
            }
        }

        private void InitArcs()
        {
            for (int arcIndex = 0; arcIndex < _wordGraph.Arcs.Count; arcIndex++)
            {
                WordGraphArc arc = _wordGraph.Arcs[arcIndex];

                // init ecm score info for each word of arc
                EcmScoreInfo prevEsi = _stateEcmScoreInfos[arc.PrevState];
                var esis = new List<EcmScoreInfo>();
                foreach (string word in arc.TargetTokens)
                {
                    var esi = new EcmScoreInfo();
                    _ecm.SetupEsi(esi, prevEsi, word);
                    esis.Add(esi);
                    prevEsi = esi;
                }
                _arcEcmScoreInfos.Add(esis);

                // init best scores for the arc's successive state
                UpdateStateBestScores(arcIndex, 0);

                _statesInvolvedInArcs.Add(arc.PrevState);
                _statesInvolvedInArcs.Add(arc.NextState);
            }
        }

        private void UpdateInitialStateBestScores()
        {
            EcmScoreInfo esi = _stateEcmScoreInfos[WordGraph.InitialState];

            _stateWordGraphScores[WordGraph.InitialState] = _wordGraph.InitialStateScore;

            List<double> bestScores = _stateBestScores[WordGraph.InitialState];
            List<int> bestPrevArcs = _stateBestPrevArcs[WordGraph.InitialState];

            bestScores.Clear();
            bestPrevArcs.Clear();

            foreach (double score in esi.Scores)
            {
                bestScores.Add((EcmWeight * -score) + (WordGraphWeight * _wordGraph.InitialStateScore));
                bestPrevArcs.Add(int.MaxValue);
            }
        }

        private void UpdateStateBestScores(int arcIndex, int prefixDiffSize)
        {
            WordGraphArc arc = _wordGraph.Arcs[arcIndex];
            List<EcmScoreInfo> arcEsis = _arcEcmScoreInfos[arcIndex];

            EcmScoreInfo prevEsi = arcEsis.Count == 0 ? _stateEcmScoreInfos[arc.PrevState] : arcEsis[arcEsis.Count - 1];

            double wordGraphScore = _stateWordGraphScores[arc.PrevState] + arc.Score;

            List<double> nextStateBestScores = _stateBestScores[arc.NextState];
            List<int> nextStateBestPrevArcs = _stateBestPrevArcs[arc.NextState];

            var positions = new List<int>();
            int startPos = prefixDiffSize == 0 ? 0 : prevEsi.Scores.Count - prefixDiffSize;
            for (int i = startPos; i < prevEsi.Scores.Count; i++)
            {
                double newScore = (EcmWeight * -prevEsi.Scores[i]) + (WordGraphWeight * wordGraphScore);

                if (i == nextStateBestScores.Count || nextStateBestScores[i] < newScore)
                {
                    AddOrReplace(nextStateBestScores, i, newScore);
                    positions.Add(i);
                    AddOrReplace(nextStateBestPrevArcs, i, arcIndex);
                }
            }

            _stateEcmScoreInfos[arc.NextState].UpdatePositions(prevEsi, positions);

            if (wordGraphScore > _stateWordGraphScores[arc.NextState])
                _stateWordGraphScores[arc.NextState] = wordGraphScore;
        }

        private void AddOrReplace<T>(List<T> list, int index, T item)
        {
            if (index > list.Count)
                throw new ArgumentOutOfRangeException(nameof(index));

            if (index == list.Count)
                list.Add(item);
            else
                list[index] = item;
        }

        public double EcmWeight { get; }
        public double WordGraphWeight { get; }

        public void Correct(IReadOnlyList<string> prefix, bool isLastWordComplete)
        {
            // get valid portion of the processed prefix vector
            int validProcPrefixCount = 0;
            for (int i = 0; i < _prevPrefix.Length; i++)
            {
                if (i >= prefix.Count)
                    break;

                if (i == _prevPrefix.Length - 1 && i == prefix.Count - 1)
                {
                    if (_prevPrefix[i] == prefix[i] && _prevIsLastWordComplete == isLastWordComplete)
                        validProcPrefixCount++;
                }
                else if (_prevPrefix[i] == prefix[i])
                {
                    validProcPrefixCount++;
                }
            }

            int diffSize = _prevPrefix.Length - validProcPrefixCount;
            if (diffSize > 0)
            {
                // adjust size of info for arcs
                foreach (List<EcmScoreInfo> esis in _arcEcmScoreInfos)
                {
                    foreach (EcmScoreInfo esi in esis)
                    {
                        for (int i = 0; i < diffSize; i++)
                            esi.RemoveLast();
                    }
                }

                // adjust size of info for states
                foreach (int state in _statesInvolvedInArcs)
                {
                    for (int i = 0; i < diffSize; i++)
                    {
                        _stateEcmScoreInfos[state].RemoveLast();
                        _stateBestScores[state].RemoveAt(_stateBestScores[state].Count - 1);
                        _stateBestPrevArcs[state].RemoveAt(_stateBestPrevArcs[state].Count - 1);
                    }
                }
            }

            // get difference between prefix and valid portion of processed prefix
            var prefixDiff = new string[prefix.Count - validProcPrefixCount];
            for (int i = 0; i < prefixDiff.Length; i++)
                prefixDiff[i] = prefix[validProcPrefixCount + i];

            // process word-graph given prefix difference
            ProcessWordGraphForPrefixDiff(prefixDiff, isLastWordComplete);

            _prevPrefix = prefix.ToArray();
            _prevIsLastWordComplete = isLastWordComplete;
        }

        public IEnumerable<TranslationResult> GetResults()
        {
            PriorityQueue<Hypothesis> queue = GetHypotheses();

            foreach (Hypothesis hypothesis in Search(queue))
            {
                var builder = new TranslationResultBuilder(_wordGraph.SourceTokens)
                {
                    TargetDetokenizer = _targetDetokenizer
                };
                BuildCorrectionFromHypothesis(builder, _prevPrefix, _prevIsLastWordComplete, hypothesis);
                yield return builder.ToResult();
            }
        }

        private IEnumerable<Hypothesis> Search(PriorityQueue<Hypothesis> queue)
        {
            while (!queue.IsEmpty)
            {
                Hypothesis hypothesis = queue.Dequeue();
                int lastState =
                    hypothesis.Arcs.Count == 0
                        ? hypothesis.StartState
                        : hypothesis.Arcs[hypothesis.Arcs.Count - 1].NextState;

                if (_wordGraph.FinalStates.Contains(lastState))
                {
                    yield return hypothesis;
                }
                else if (ConfidenceThreshold <= 0)
                {
                    hypothesis.Arcs.AddRange(_wordGraph.GetBestPathFromStateToFinalState(lastState));
                    yield return hypothesis;
                }
                else
                {
                    double score = hypothesis.Score - (WordGraphWeight * _restScores[lastState]);
                    IReadOnlyList<int> arcIndices = _wordGraph.GetNextArcIndices(lastState);
                    bool enqueuedArc = false;
                    for (int i = 0; i < arcIndices.Count; i++)
                    {
                        int arcIndex = arcIndices[i];
                        WordGraphArc arc = _wordGraph.Arcs[arcIndex];
                        if (IsArcPruned(arc))
                            continue;

                        Hypothesis newHypothesis = hypothesis;
                        if (i < arcIndices.Count - 1)
                            newHypothesis = newHypothesis.Clone();
                        newHypothesis.Score = score;
                        newHypothesis.Score += arc.Score;
                        newHypothesis.Score += _restScores[arc.NextState];
                        newHypothesis.Arcs.Add(arc);
                        queue.Enqueue(newHypothesis);
                        enqueuedArc = true;
                    }

                    if (!enqueuedArc && (hypothesis.StartArcIndex != -1 || hypothesis.Arcs.Count > 0))
                    {
                        hypothesis.Arcs.AddRange(_wordGraph.GetBestPathFromStateToFinalState(lastState));
                        yield return hypothesis;
                    }
                }
            }
        }

        private void ProcessWordGraphForPrefixDiff(string[] prefixDiff, bool isLastWordComplete)
        {
            if (prefixDiff.Length == 0)
                return;

            if (!_wordGraph.IsEmpty)
            {
                EcmScoreInfo prevInitialEsi = _stateEcmScoreInfos[WordGraph.InitialState];
                _ecm.ExtendInitialEsi(_stateEcmScoreInfos[WordGraph.InitialState], prevInitialEsi, prefixDiff);
                UpdateInitialStateBestScores();
            }

            for (int arcIndex = 0; arcIndex < _wordGraph.Arcs.Count; arcIndex++)
            {
                WordGraphArc arc = _wordGraph.Arcs[arcIndex];

                // update ecm score info for each word of arc
                EcmScoreInfo prevEsi = _stateEcmScoreInfos[arc.PrevState];
                List<EcmScoreInfo> esis = _arcEcmScoreInfos[arcIndex];
                while (esis.Count < arc.TargetTokens.Count)
                    esis.Add(new EcmScoreInfo());
                for (int i = 0; i < arc.TargetTokens.Count; i++)
                {
                    EcmScoreInfo esi = esis[i];
                    _ecm.ExtendEsi(
                        esi,
                        prevEsi,
                        arc.IsUnknown ? string.Empty : arc.TargetTokens[i],
                        prefixDiff,
                        isLastWordComplete
                    );
                    prevEsi = esi;
                }

                // update best scores for the arc's successive state
                UpdateStateBestScores(arcIndex, prefixDiff.Length);
            }
        }

        private PriorityQueue<Hypothesis> GetHypotheses()
        {
            var queue = new PriorityQueue<Hypothesis>(1000);

            // add hypotheses starting before each word in each arc
            for (int arcIndex = 0; arcIndex < _wordGraph.Arcs.Count; arcIndex++)
            {
                WordGraphArc arc = _wordGraph.Arcs[arcIndex];
                if (!IsArcPruned(arc))
                {
                    double wordGraphScore = _stateWordGraphScores[arc.PrevState] + arc.Score;

                    for (int i = -1; i < arc.TargetTokens.Count - 1; i++)
                    {
                        EcmScoreInfo esi =
                            i == -1 ? _stateEcmScoreInfos[arc.PrevState] : _arcEcmScoreInfos[arcIndex][i];
                        double score =
                            (WordGraphWeight * wordGraphScore)
                            + (EcmWeight * -esi.Scores[esi.Scores.Count - 1])
                            + (WordGraphWeight * _restScores[arc.NextState]);
                        queue.Enqueue(new Hypothesis(score, arc.NextState, arcIndex, i));
                    }
                }
            }

            // add hypotheses starting before each final state
            foreach (int state in _wordGraph.FinalStates)
            {
                double restScore = _restScores[state];
                List<double> bestScores = _stateBestScores[state];

                double score = bestScores[bestScores.Count - 1] + (WordGraphWeight * restScore);
                queue.Enqueue(new Hypothesis(score, state));
            }

            return queue;
        }

        private bool IsArcPruned(WordGraphArc arc)
        {
            return !arc.IsUnknown && arc.Confidences.Any(c => c < ConfidenceThreshold);
        }

        private void BuildCorrectionFromHypothesis(
            TranslationResultBuilder builder,
            string[] prefix,
            bool isLastWordComplete,
            Hypothesis hypothesis
        )
        {
            int uncorrectedPrefixLen;
            if (hypothesis.StartArcIndex == -1)
            {
                AddBestUncorrectedPrefixState(builder, prefix.Length, hypothesis.StartState);
                uncorrectedPrefixLen = builder.TargetTokens.Count;
            }
            else
            {
                AddBestUncorrectedPrefixSubState(
                    builder,
                    prefix.Length,
                    hypothesis.StartArcIndex,
                    hypothesis.StartArcWordIndex
                );
                WordGraphArc firstArc = _wordGraph.Arcs[hypothesis.StartArcIndex];
                uncorrectedPrefixLen =
                    builder.TargetTokens.Count - (firstArc.TargetTokens.Count - hypothesis.StartArcWordIndex) + 1;
            }

            int alignmentColsToAddCount = _ecm.CorrectPrefix(builder, uncorrectedPrefixLen, prefix, isLastWordComplete);

            foreach (WordGraphArc arc in hypothesis.Arcs)
            {
                UpdateCorrectionFromArc(builder, arc, alignmentColsToAddCount);
                alignmentColsToAddCount = 0;
            }
        }

        private void AddBestUncorrectedPrefixState(TranslationResultBuilder builder, int procPrefixPos, int state)
        {
            var arcs = new Stack<WordGraphArc>();

            int curState = state;
            int curProcPrefixPos = procPrefixPos;
            while (curState != WordGraph.InitialState)
            {
                int arcIndex = _stateBestPrevArcs[curState][curProcPrefixPos];
                WordGraphArc arc = _wordGraph.Arcs[arcIndex];

                for (int i = arc.TargetTokens.Count - 1; i >= 0; i--)
                {
                    IReadOnlyList<int> predPrefixWords = _arcEcmScoreInfos[arcIndex][i].GetLastInsPrefixWordFromEsi();
                    curProcPrefixPos = predPrefixWords[curProcPrefixPos];
                }

                arcs.Push(arc);

                curState = arc.PrevState;
            }

            foreach (WordGraphArc arc in arcs)
                UpdateCorrectionFromArc(builder, arc, 0);
        }

        private void AddBestUncorrectedPrefixSubState(
            TranslationResultBuilder builder,
            int procPrefixPos,
            int arcIndex,
            int arcWordIndex
        )
        {
            WordGraphArc arc = _wordGraph.Arcs[arcIndex];

            int curProcPrefixPos = procPrefixPos;
            for (int i = arcWordIndex; i >= 0; i--)
            {
                IReadOnlyList<int> predPrefixWords = _arcEcmScoreInfos[arcIndex][i].GetLastInsPrefixWordFromEsi();
                curProcPrefixPos = predPrefixWords[curProcPrefixPos];
            }

            AddBestUncorrectedPrefixState(builder, curProcPrefixPos, arc.PrevState);

            UpdateCorrectionFromArc(builder, arc, 0);
        }

        private void UpdateCorrectionFromArc(
            TranslationResultBuilder builder,
            WordGraphArc arc,
            int alignmentColsToAddCount
        )
        {
            for (int i = 0; i < arc.TargetTokens.Count; i++)
                builder.AppendToken(arc.TargetTokens[i], arc.Sources[i], arc.Confidences[i]);

            WordAlignmentMatrix alignment = arc.Alignment;
            if (alignmentColsToAddCount > 0)
            {
                var newAlignment = new WordAlignmentMatrix(
                    alignment.RowCount,
                    alignment.ColumnCount + alignmentColsToAddCount
                );
                for (int j = 0; j < alignment.ColumnCount; j++)
                {
                    for (int i = 0; i < alignment.RowCount; i++)
                        newAlignment[i, alignmentColsToAddCount + j] = alignment[i, j];
                }
                alignment = newAlignment;
            }

            builder.MarkPhrase(arc.SourceSegmentRange, alignment);
        }

        private class Hypothesis : PriorityQueueNodeBase, IComparable<Hypothesis>
        {
            public Hypothesis(double score, int startState, int startArcIndex = -1, int startArcWordIndex = -1)
            {
                Score = score;
                StartState = startState;
                StartArcIndex = startArcIndex;
                StartArcWordIndex = startArcWordIndex;
                Arcs = new List<WordGraphArc>();
            }

            public Hypothesis(Hypothesis other)
            {
                Score = other.Score;
                StartState = other.StartState;
                StartArcIndex = other.StartArcIndex;
                StartArcWordIndex = other.StartArcWordIndex;
                Arcs = other.Arcs.ToList();
            }

            public double Score { get; set; }
            public int StartState { get; }
            public int StartArcIndex { get; }
            public int StartArcWordIndex { get; }
            public List<WordGraphArc> Arcs { get; }

            public Hypothesis Clone()
            {
                return new Hypothesis(this);
            }

            public int CompareTo(Hypothesis other)
            {
                return -Score.CompareTo(other.Score);
            }
        }
    }
}
